junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/UniqueIdTrackingListener.java

```diff
@@ -138,22 +139,37 @@ public UniqueIdTrackingListener() {
 	@Override
 	public void testPlanExecutionStarted(TestPlan testPlan) {
 		this.enabled = testPlan.getConfigurationParameters().getBoolean(LISTENER_ENABLED_PROPERTY_NAME).orElse(false);
+		this.testPlan = testPlan;
 	}
 
 	@Override
 	public void executionSkipped(TestIdentifier testIdentifier, String reason) {
-		trackTestUid(testIdentifier);
+		if (this.enabled) {
+			// When a container is skipped, there are no events for its children.
+			// Therefore, in order to track them, we need to traverse the subtree.
+			trackTestUidRecursively(testIdentifier);
+		}
 	}
 
 	@Override
 	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
-		trackTestUid(testIdentifier);
+		if (this.enabled) {
+			trackTestUid(testIdentifier);
+		}
+	}
+
+	private void trackTestUidRecursively(TestIdentifier testIdentifier) {
+		boolean tracked = trackTestUid(testIdentifier);
+		if (!tracked) {
+			this.testPlan.getChildren(testIdentifier).forEach(this::trackTestUidRecursively);
+		}
 	}
 
-	private void trackTestUid(TestIdentifier testIdentifier) {
-		if (this.enabled && testIdentifier.isTest()) {
-			this.uniqueIds.add(testIdentifier.getUniqueId());
+	private boolean trackTestUid(TestIdentifier testIdentifier) {
+		if (testIdentifier.isTest()) {
+			return this.uniqueIds.add(testIdentifier.getUniqueId());
```

Comment:
This will always return `true` since it's an `ArrayList`.

Did you perhaps mean to switch to a `LinkedHashSet`?

If not, perhaps we should change that to the following for better clarity.

```suggestion
			this.uniqueIds.add(testIdentifier.getUniqueId());
			return true;
```

Reply:
I know that it will always return `true`. I kept the `ArrayList` to avoid changing to much. That being said, I think your suggestion does make it clearer so I'll follow it.